File: Ashutosh-Singh/ClassQuestions/8-JAN-25/SpiralMatrix-2.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> matrix(n, vector<int>(n, 1));

        int count = 0;
        int total = n * n;
        int value = 1;
        // indexes
        int left = 0;
        int right = n - 1;
        int top = 0;
        int down = n - 1;

        while (count < total) {
            // firstRow
            for (int i = left; count < total && i <= right; i++) {
                matrix[top][i] = value;
                count++;
                value++;
            }
            top++;

            // endColumn
            for (int i = top; count < total && i <= down; i++) {
                matrix[i][right] = value;
                count++;
                value++;
            }
            right--;

            // lastRow
            for (int i = right; count < total && i >= left; i--) {
                matrix[down][i] = value;
                count++;
                value++;
            }
            down--;

            // firstCol
            for (int i = down; count < total && i >= top; i--) {
                matrix[i][left] = value;
                count++;
                value++;
            }
            left++;
        }
        return matrix;
    }
};

int main() {
    int n;
    cout << "Enter the size of the matrix (n): ";
    cin >> n;

    Solution solution;
    vector<vector<int>> result = solution.generateMatrix(n);

    cout << "The generated matrix is:" << endl;
    for (const auto& row : result) {
        for (int num : row) {
            cout << num << " ";
        }
        cout << endl;
    }

    return 0;
}

File: [Manya]/Leetcode/2D-Array/Spiral_matrix_2.java
import java.util.*;

class Solution {
    public int[][] generateMatrix(int n) {
        int r1 = 0;
        int c1 = 0;
        int r2 = n - 1;
        int c2 = n - 1;
        int[][] answer = new int[n][n];
        int nums = 1;

        while (r1 <= r2 && c1 <= c2) {
            for (int c = c1; c <= c2; c++) {
                answer[r1][c] = nums++;
            }
            for (int r = r1 + 1; r <= r2; r++) {
                answer[r][c2] = nums++;
            }
            if (r1 < r2 && c1 < c2) {
                for (int c = c2 - 1; c >= c1; c--) {
                    answer[r2][c] = nums++;
                }
                for (int r = r2 - 1; r > r1; r--) {
                    answer[r][c1] = nums++;
                }
            }
            r1++;
            c1++;
            r2--;
            c2--;
        }
        return answer;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Enter the size of the matrix (n): ");
        int n = scanner.nextInt();
        
        Solution solution = new Solution();
        int[][] result = solution.generateMatrix(n);
        
        System.out.println("Generated Spiral Matrix:");
        for (int[] row : result) {
            System.out.println(Arrays.toString(row));
        }
        
        scanner.close();
    }
}